### ISSUE TITLE
[help] XML Builder and Newlines Oddity

### ISSUE THREAD

**Issue by svoop** [sparklemotion/nokogiri#2522]
To build large and rather cumbersome AIXM (XML) files, I'm currently switching from the "builder" gem to Nokogiri's built-in XML builder. `DocumentFragment`s come in quite handy to build the blocks with which to assemble the final document.

Most `DocumentFragment`s have a container with which everything works like a charm:

```ruby
require 'nokogiri'

Nokogiri::XML::DocumentFragment.parse('').tap do |document|
  Nokogiri::XML::Builder.with(document) do |builder|
    builder.root do |root|
      root.foo('bar')
      root.fii('bir')
    end
  end
end.to_xml
```

```xml
<root>
  <foo>bar</foo>
  <fii>bir</fii>
</root>
```

Unfortunately, a few `DocumentFragment`s merely contain a bunch of elements. For some reason, the result is not formatted anymore, everything ends up on one line:

```ruby
require 'nokogiri'

Nokogiri::XML::DocumentFragment.parse('').tap do |document|
  Nokogiri::XML::Builder.with(document) do |builder|
    builder.foo('bar')
    builder.fii('bir')
  end
end.to_xml
```

```xml
<foo>bar</foo><fii>bir</fii>
```

Is there a way have newslines after each element as in the containered version?

(I've unsuccessfully looked for methods to extract part of a document into a new `DocumentFragment`, say, using the first example to build, then extract a `DocumentFragment` containing only the children of `root`.)

Thanks for your help!

**Comment by svoop** [comment]
From what I read about other implementations using libxml2, this appears to be a "feature" of libxml2 rather than to happen on Nokogiri's end, right?

In any case, I'm working around this readability issue likes so now:

```ruby
def build_fragment
  Nokogiri::XML::DocumentFragment.parse('').tap do |document|
    Nokogiri::XML::Builder.with(document) do |builder|
      yield builder
    end
    document.elements.each { _1.add_next_sibling("\n") }   # add newline between tags on top level
  end
end
```

Then use it:

```ruby
build_fragment do |builder|
  builder.foo('bar')
  builder.fii('bir')
end
```

To get:

```xml
<foo>bar</foo>
<fii>bir</fii>
```

**Comment by flavorjones** [comment]
Please see my explanation at https://github.com/sparklemotion/nokogiri/issues/2521#issuecomment-1114018266 for how libxml2 decides to format subtrees when serializing.

TL;DR, if any child of a node is a TEXT, CDATA, or ENTITY_REF node, then that node (and recursively its subtree) will be printed literally, without formatting and indentation.

**Comment by svoop** [comment]
I came across this bit somewhere deep down on Stackoverflow. And knowing that this decision is made by libxml2 ultimately led to this workaround. It's nicely tucked away and doesn't pollute my builder models, so works for me. 🎉
